Implements tests for Double Matchers

Kerooker opened the pull request:
Given the necessity to continually improve the code, having unit tests guarantees that nothing will break when a change is made. This commit focuses on all Double Matchers, refactoring their tests to cover a lot more scenarios and corner cases.

One problem that this commit introduces is that there are way too many lines for a single class. This exposes a code smell on the matchers.kt file, in which ALL the double matchers are placed. This is bad, as each responsibility should be refactored to it's own class. This correction is possible when issue #450 is solved.

Solves https://github.com/kotlintest/kotlintest/issues/430

Kerooker commented:
@sksamuel and @ajalt 

Could you help me with this by code-reviewing?

I know this class is massive right now, but it will test every possible scenario. With a possible refactor from #450, we will get this into a smaller size.

However, #450 is only possible with correct tests for the classes, which are in this PR.

ajalt commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1061 @@
+          
+          "Any number smaller than itself" {
+            assertAll(nonMaxValueDoubles) {
+              it shouldNotMatchExactly it - delta
```
Tests like this will only pass for values with an exponent relatively close to 1. For example:
`1e300 + 0.0001 === 1e300`
`1e300 + 0.0001 === 0.0001`

Kerooker replied:
I didn't quite understand what you meant.

Do you believe this test may sometimes fail?

ajalt replied:
It looks like you're trying to create a number `it - 0.0001` that is different from `it`. But for most `Double` values, `it - 0.0001` is exactly equal to `it`.

[Double precision floating point](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) can only represent about 16 significant decimal digits. `1e300 - 0.0001` would require 300 significant decimal digits, so the result just rounds back to the original `1e300`.

So your delta needs to be relative to the value you're trying to change. For example, rather than `it - delta`, you could do something like `it * .9`, which should work for all numbers with a magnitude larger than ~1e-323 (the smallest positive denormal number).

Kerooker replied:
That's very nice to know! Learn something every day, I guess.

I'll add these corner case numbers to the generated double lists

ajalt commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1061 @@
+    }
+  }
+  
+  private infix fun Double.shouldExactlyMatch(other: Double) {
```
I like these functions that increase coverage without duplication 👍

Kerooker replied:
They're good because we expect the exact same behavior now. But this is dangerous, because if I change any behavior in the future, the old tests wouldn't cover them. This way we can cover on all fronts, including trying the exception ones

ajalt commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1061 @@
+          "When it's equal to the first number of the range" - {
+            
+            "With tolerance" {
+              onDoubleRange { double, _ ->
```
If you aren't using `valueToVariate`, then this is taking 10x the time without really helping. I would suggest useing `assertAll` directly in these cases.

Kerooker replied:
Good call

ajalt commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1061 @@
+          "Numbers smaller than itself" {
+            onDoubleRange { double, valueToVariate ->
+              double shouldNotMatchLessThan double - delta
+              double shouldNotMatchLessThan double - valueToVariate
```
I have the same comment about `valueToVariate` that I did about `delta`: it's not valid if `double` has a large positive or negative exponent.

Kerooker replied:
I'm not quite sure if I understood this. Do you have suggestions on how to deal with this?

sksamuel commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1061 @@
+                              { this shouldNotBe between(a, b, tolerance) })
+  }
+  
+  private fun onDoubleRange(block: (double: Double, valueToVariate: Double) -> Unit) {
```
Are you using the valueToVariate parameter here ?

sksamuel commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1061 @@
+  
+  private infix fun ClosedRange<Double>.step(step: Double): Iterable<Double> {
+    val sequence = generateSequence(start) { previous ->
+      if (previous == Double.POSITIVE_INFINITY) return@generateSequence null
```
I think return@ is a code smell. It's more functional to have the return be the last value all the time. I'd simply change this to have the next two lines inside an else {}

sksamuel commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1061 @@
+  
+  private infix fun Double.shouldNotMatchLessThan(x: Double) {
+    this shouldNotBe lt(x)
+    this shouldNotBeLessThan x
```
If this fails, won't the next two lines not be executed?

Kerooker replied:
That's correct, but I don't think it's a bad thing. We want the test to fail when this fails

ajalt replied:
We have `assertSoftly` for this case.

Kerooker commented:
@sksamuel and @ajalt 
Take a look now, I addressed your issues.

However, @sksamuel, Trevis is unable to test anything for quite some time. I think something broke ./gradlew test after [this commit](https://github.com/kotlintest/kotlintest/commit/74c132ebdb72b99a93f647cc1fbb8214f01eb4c6)

ajalt commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1085 @@
+    
+  }
+  
+  private fun Double.smallDelta(): Double {
```
We're getting there! These extensions work in more cases, but still fail for values near the representable limits. For example, `smallDelta` and the `slighty*` extensions fail because `Double.MIN_VALUE * 0.01 == Double.MIN_VALUE`. You're looking for the smallest number you can add to a double that will change its value. The term for this is [Unit of Least Precision (ULP)](https://en.wikipedia.org/wiki/Unit_in_the_last_place). The JDK has a built in function for this: [`Math.ulp`](https://docs.oracle.com/javase/10/docs/api/java/lang/Math.html#ulp(double)).

For the `muchGreater` and `muchSmaller` extensions, many large double values will end up as an infinity (e.g. `1.5e308 * 1.25 == POSITIVE_INFINITY`), and many small values will produce incorrect results (e.g. `3e-324.muchGreater() == Double.MIN_VALUE`, which is _smaller_ than `3e-324`). I suggest using a multiple of the ULP like `this + 2 * Math.ulp(this)`.

Kerooker replied:
Incredible!! I'm learning a lot with this! Will fix this by tomorrow.

Thanks a lot for the insight!

Kerooker replied:
I think this is fixed now.

ajalt commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1065 @@
+  
+  
+  private fun Double.toleranceValue(): Double {
+    return 5 * ulp
```
This is fine, but I would use `ulp, `2 * ulp`, and `3 * ulp` rather that 5, 10, and 100. The larger values don't catch any extra errors, and will cause more values to be rounded to infinity.

ajalt commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1065 @@
+  }
+  
+  private fun Double.slightlyGreater(): Double {
+    return this + (10 *ulp)
```
Can you reformat the file? The spacing is a little messed up here.

ajalt commented on kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt:
```diff
@@ -1,72 +1,1065 @@
+  }
+  
+  
+  private fun Double.shouldThrowExceptionOnBetween(a: Double,
```
This is an extremely minor nit, but the kotlin style guide recommends formatting multi-line parameters like this:

```kotlin
private fun Double.shouldThrowExceptionOnBetween(
    a: Double,
    b: Double,
    tolerance: Double,
    message: String = "$this should be smaller than $b and bigger than $a"
) {
```

Which is nice in this case to avoid the long line due to the default parameter.

Kerooker commented:
I agree to wait for CI. The issue for it is https://github.com/kotlintest/kotlintest/issues/457, if you'd like to follow it
